Gems/AtomTressFX/Code/Builders/HairAssetBuilder.cpp

```diff
@@ -0,0 +1,168 @@
+                AssetBuilderSDK::SourceFileDependency sourceFileDependency;
+                sourceFileDependency.m_sourceFileDependencyPath = request.m_sourceFile;
+                AZ::StringFunc::Path::ReplaceExtension(sourceFileDependency.m_sourceFileDependencyPath, AMD::TFXBoneFileExtension);
+                response.m_sourceFileDependencyList.push_back(sourceFileDependency);
+
+                sourceFileDependency.m_sourceFileDependencyPath = request.m_sourceFile;
+                AZ::StringFunc::Path::ReplaceExtension(sourceFileDependency.m_sourceFileDependencyPath, AMD::TFXMeshFileExtension);
+                response.m_sourceFileDependencyList.push_back(sourceFileDependency);
+
+                response.m_result = AssetBuilderSDK::CreateJobsResultCode::Success;
+            }
+
+            void HairAssetBuilder::ProcessJob(
+                const AssetBuilderSDK::ProcessJobRequest& request, AssetBuilderSDK::ProcessJobResponse& response)
+            {
+                AZ_TracePrintf(AssetBuilderSDK::InfoWindow, "HairAssetBuilder Starting Job.\n");
+
+                if (m_isShuttingDown)
+                {
+                    AZ_TracePrintf(
+                        AssetBuilderSDK::WarningWindow, "Cancelled job %s because shutdown was requested.\n", request.m_fullPath.c_str());
+                    response.m_resultCode = AssetBuilderSDK::ProcessJobResult_Cancelled;
+                    return;
+                }
+
+                // There are 3 source files for TressFX asset - .tfx, .tfxbone and .tfxmesh.
+                // We read all the 3 source files and combine them to 1 output file .tfxhair in the cache.
+                AZStd::string tfxFileName;
+                AzFramework::StringFunc::Path::GetFullFileName(request.m_fullPath.c_str(), tfxFileName);
+
+                // Create the path to the result .tfxhair file inside the tempDirPath.
+                AZStd::string destPath;
+                AzFramework::StringFunc::Path::ConstructFull(request.m_tempDirPath.c_str(), tfxFileName.c_str(), destPath, true);
+                AZ::StringFunc::Path::ReplaceExtension(destPath, AMD::TFXCombinedFileExtension);
+
+                // Create and open the .tfxhair we are writing to. 
+                AZ::IO::FileIOStream outStream(destPath.data(), AZ::IO::OpenMode::ModeWrite | AZ::IO::OpenMode::ModeCreatePath);
+                if (!outStream.IsOpen())
+                {
+                    AZ_TracePrintf(
+                        AssetBuilderSDK::ErrorWindow, "Error: Failed job %s because .tfxhair file cannot be created.\n", request.m_fullPath.c_str());
+                    response.m_resultCode = AssetBuilderSDK::ProcessJobResult_Failed;
+                    return;
+                }
+
+                // Write the .tfxhair file header as a placeholder
+                AMD::TressFXCombinedHairFileHeader header; 
+                outStream.Write(sizeof(AMD::TressFXCombinedHairFileHeader), &header);
```

Comment:
This is a good question, and I believe it is because the Asset is really a wrapper around 3 headers that are stored together and represent a single asset.   The three headers are then used for generating the streams and storing for the actual loaders.
Can we do it differently?  perhaps, although there is no reason why. Since Roman implemented this piece, I can refer this to him if you feel that this is important enough.

Reply:
The header file contain information about output file. It's writing an empty header as a placeholder here to make sure the default value has been written to the head.
An alternative world be using a seek instead, because it basically just reserves the header. As long as the actual header will be written later, it would be safe.